File: Sources/TamaCore/Pet/CatStateStore.swift
import Foundation

/// Persists the cat's own `CatState` as JSON. The directory is injected (real app →
/// Application Support; tests → a temp dir). This is the app's ONLY write, and it lands
/// strictly inside the injected directory — never in any agent log tree. Missing or
/// malformed files fall back to `.initial` and never crash.
public struct CatStateStore {
    private let fileURL: URL
    private let fileManager: FileManager

    public init(directory: URL, fileManager: FileManager = .default) {
        self.fileURL = directory.appendingPathComponent("cat-state.json")
        self.fileManager = fileManager
    }

    public func load() -> CatState {
        guard let data = try? Data(contentsOf: fileURL),
              let state = try? JSONDecoder().decode(CatState.self, from: data) else {
            return .initial
        }
        return state
    }

    public func save(_ state: CatState) {
        guard let data = try? JSONEncoder().encode(state) else { return }
        try? fileManager.createDirectory(at: fileURL.deletingLastPathComponent(),
                                         withIntermediateDirectories: true)
        try? data.write(to: fileURL, options: .atomic)
    }

    /// The shipped store: `~/Library/Application Support/Tama/`. Falls back to the temporary
    /// directory if Application Support can't be resolved, so it never force-unwraps.
    public static func applicationSupport(fileManager: FileManager = .default) -> CatStateStore {
        let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? fileManager.temporaryDirectory
        return CatStateStore(directory: base.appendingPathComponent("Tama", isDirectory: true),
                             fileManager: fileManager)
    }
}

File: windows/Tama.Core/CatStateStore.cs
using System.Text.Json;

namespace Tama.Core;

/// <summary>
/// Persists the cat's own CatState as JSON in the injected directory (real app → %APPDATA%\Tama;
/// tests → a temp dir). This is the app's ONLY write, strictly inside that directory — never in
/// any agent log tree. Missing/malformed files fall back to Initial and never crash.
/// </summary>
public sealed class CatStateStore
{
    private readonly string _filePath;

    public CatStateStore(string directory) =>
        _filePath = Path.Combine(directory, "cat-state.json");

    public CatState Load()
    {
        try
        {
            var doc = JsonDocument.Parse(File.ReadAllBytes(_filePath));
            using var _ = doc;
            if (doc.RootElement.ValueKind != JsonValueKind.Object) return CatState.Initial;
            if (doc.RootElement.TryGetProperty("lastSeenDay", out var v)
                && v.ValueKind == JsonValueKind.String
                && DateTimeOffset.TryParse(v.GetString(), out var d))
                return new CatState(d);
            return CatState.Initial;
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException or JsonException
            or ArgumentException or NotSupportedException)
        { return CatState.Initial; }
    }

    public void Save(CatState state)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_filePath)!);
            var json = JsonSerializer.Serialize(new
            {
                lastSeenDay = state.LastSeenDay?.ToString("O"),
            });
            var tmp = _filePath + ".tmp";
            File.WriteAllText(tmp, json);
            File.Move(tmp, _filePath, overwrite: true);   // atomic-ish replace
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException
            or ArgumentException or NotSupportedException)
        { /* persistence is best-effort; mood still works from Initial next launch */ }
    }

    /// <summary>The shipped store: %APPDATA%\Tama (falls back to the temp dir, never throws).</summary>
    public static CatStateStore AppData()
    {
        var basePath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        if (string.IsNullOrEmpty(basePath)) basePath = Path.GetTempPath();
        return new CatStateStore(Path.Combine(basePath, "Tama"));
    }
}
